### PR TITLE
Suggestion: Panic on loops in the functional reactive data flow

### PR DESCRIPTION
**Pull request by s9ferech**
### Pull Request Description

In the current implementation, loops in the data flow of a functional reactive network are detected and the data flow is quietly interrupted at that point to prevent infinite looping. Such an interruption would likely lead to behaviour that is hard to predict. Panicking instead could make it much easier to identify underlying bugs.

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md) style guide.
- [ ] All code has been tested where possible.
- [ ] All code has been profiled where possible.

### CODE REVIEW

**Comment by wdanilo** [comment]
Thank you for the suggestion! There is a problem with it though, namely, the design which breaks loops was done on purpose, is very popular in FRP implementations. However, it is true that we have never used it for a real world use-case. Moreover, I always prefer warnings than crashing the gui, so I'd suggest to print warning about it to console instead.

**Comment by s9ferech** [comment]
Thanks for the answer. Out of curiosity, do you have a resource explaining the motivations behind this kind of design? Or could you give me a brief overview?

**Comment by wdanilo** [comment]
Oh, the reason is pretty simple here - basically, you prefer your software to recover on production rather than crash. As FRP systems are used in many "reactive behaviours", like GUI, or device signal processing, you prefer your application to still work on client-side rather than break at some point. So collecting logs and fixing that always looks better than software that crashesh from both user nad business perspective. This is exactly the same reason why JS is used more and more to code GUI for games on PC and consoles - it is VERY forgiving. So there is higher probability that user will see "undefined" somewhere in the game rather than game crashing, which makes the final product just more stable. Does it make sense?

**Comment by wdanilo** [comment]
Btw, could you use our `logger` functionality to report warnings here, please?

**Comment by s9ferech** [comment]
In principle, it makes sense. I just don't have enough experience and knowledge of the Enso IDE yet to really understand this kind of trade-off.

**Comment by s9ferech** [comment]
> Btw, could you use our `logger` functionality to report warnings here, please?

Do you mean the GitHub issue tracker?

**Comment by s9ferech** [comment]
> Btw, could you use our `logger` functionality to report warnings here, please?

Which verbosity level would be appropriate?

**Comment by wdanilo** [comment]
Definitely warning!

**Comment by wdanilo** [comment]
Hey @s9ferech, are you still working on this issue? :)

**Comment by s9ferech** [comment]
I am planning to get to it. But maybe someone else should just do it since I don't understand yet how you handle logging.

File: src/rust/lib/frp/src/stream.rs [hunk 1]
```diff
@@ -175,19 +175,26 @@ impl<Out:Data> HasOutput for NodeData<Out> {
...
-            }
-            self.targets.borrow_mut().retain(|target| target.data.on_event_if_exists(value));
-            let mut new_targets = self.new_targets.borrow_mut();
-            if !new_targets.is_empty() {
-                let new_targets_ref : &mut Vec<EventInput<Out>> = &mut new_targets;
-                self.targets.borrow_mut().extend(mem::take(new_targets_ref));
-            }
-            self.during_call.set(false);
+        if self.during_call.get() {
+            let logger = Logger::new("stream::NodeData");
```
Comment (wdanilo) [inline src/rust/lib/frp/src/stream.rs:179]: The logger name is used to navigate the user logically to the place where the error occured. I'd name the logger just `frp` in this case, as using internal structure names would not tell the user a lot.

File: src/rust/lib/frp/src/stream.rs [hunk 2]
```diff
@@ -175,19 +175,26 @@ impl<Out:Data> HasOutput for NodeData<Out> {
...
+            let log_message = format!(
+                "Encountered a loop in the reactive dataflow at '{}'",
+                self.label);
+            logger.warning(&*log_message);
```
Comment (wdanilo) [inline src/rust/lib/frp/src/stream.rs:183]: You could also use our macro: `warning!(&logger, "Encountered a loop in the reactive dataflow at '{self.label}'"` :)

File: src/rust/lib/frp/src/stream.rs [hunk 3]
```diff
@@ -175,19 +175,26 @@ impl<Out:Data> HasOutput for NodeData<Out> {
...
+            return;
+        }
```
Comment (wdanilo) [inline src/rust/lib/frp/src/stream.rs:185]: We have a rule, that we try not to use early function return if we are not forced to. It just increases the brain power to understand all exit points of a function. Thus, I'd suggest to use simple `else` statement here instead.

**Review by wdanilo — COMMENTED** [review]
Amazing to see it implemented, thank you ❤️ I added a few minor comments - either apply them or reply if you dont agree with any of them. And then lets merge it! :)

**Review by wdanilo — APPROVED** [review]

**Comment by wdanilo** [comment]
I'm happy to see this implemented. Small, but important improvement, good job! ❤️

**Pull request merged**